scipy/signal/spectral.py

```diff
@@ -507,6 +507,11 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
     >>> plt.xlabel('Time [sec]')
     >>> plt.show()
     """
+    #parse window; if array like, then set nperseg = win.shape
+    if isinstance(window, string_types) or type(window) is tuple:
+        if nperseg == None: # if nperseg not specified by user or calling function
+            nperseg = 256 # change to default
+            
```

Comment:
I think the logic here is incomplete. I can call for example `spectrogram(..., nperseg=None, window=np.ones(100), noverlap=None)`. A couple lines below there is a `if noverlap is None: noverlap = nperseg // 8`, but based on your changes `nperseg` will still be `None`. So something needs to change, probably an `elif nperseg is None` to pair with your `if isinstance(...)` line.

But really the cleaner thing might be to move the window triaging up, and nest the `nperseg is None` checks in there.